Source/monster.cpp

```diff
@@ -1294,28 +1297,23 @@ void MonsterAttackMonster(int i, int mid, int hper, int mind, int maxd)
 	auto &monster = Monsters[mid];
 	assert(monster.MType != nullptr);
 
-	if (monster._mhitpoints >> 6 > 0 && (monster.MType->mtype != MT_ILLWEAV || monster._mgoal != MGOAL_RETREAT)) {
+	if (!monster.IsPossibleToHit()) {
 		int hit = GenerateRnd(100);
 		if (monster._mmode == MonsterMode::Petrified)
 			hit = 0;
-		bool unused;
-		if (!CheckMonsterHit(monster, &unused) && hit < hper) {
+		if (monster.TryLiftGargoyle())
+			return;
+		if (hit < hper) {
 			int dam = (mind + GenerateRnd(maxd - mind + 1)) << 6;
 			monster._mhitpoints -= dam;
 			if (monster._mhitpoints >> 6 <= 0) {
-				if (monster._mmode == MonsterMode::Petrified) {
-					StartDeathFromMonster(i, mid);
+				StartDeathFromMonster(i, mid);
+				if (monster._mmode == MonsterMode::Petrified)

```

Comment:
This will always fail as `StartDeathFromMonster()` sets `_mmode = Death`. If you want to avoid the duplicate `StartDeathFromMonster()` line then use `bool wasPetrified = monster._mmode == MonsterMode::Petrified;` before, or change `StartDeathFromMonster()` to not change `_mmode` if it was `Petrified`

Reply:
You are probably right.
I didn't want to go that far into monster logic in this PR. 
I noticed similiar places where this logic was confoluted with else ifs and it was previosusly looking like that:
```
//simplified pseudocode

if (petrified)
    mosterDie();
    petrify();
else
    monsterDie();
    
```
this is equivalent to
```
monsterDie();
if petrified
   Petrify(); 
```
the same was for *pseudoMonsterHit* with additions to curly braces.
I will leave the rest of this logic to monster refactoring.
I only refactored obvious cases. When things go deep into animations - it is not yet analyzed well by me.
I will try to fix places where I messed up, and try do not touch it anymore in this PR.

Reply:
Even better would be to change StartDeathFromMonster() to not call `NewMonsterAnim()` and set `_mmode = MonsterMode::Death` since then you do not need to call `monster.Petrify();` anywhere any longer.

Reply:
```cpp
void monsterDie() {petrified = false; hp = 0;}
```

Thus the two examples are not the same
```cpp
if (petrified) { // This might be true
    monsterDie();
    petrify();
} else {
    monsterDie();
    
}
```

```cpp
monsterDie();
if (petrified) { // This will never be true
   Petrify(); 
}
```

Reply:
I checked this and agree - I will just remove it it, as `StartDeathFromMonster` is always setting this, so that checks after calling it from `MonsterAttackMonster` are redundant too :| sorry for too long explanation.

Reply:
Yes I was wrong, I fixed that to bool wasPetrified (commit soon) and I call Petrify at the end. I don't want to look at what else NewMonsterAnim() is doing. Seems many things, it is safer for me to keep it out of scope and not look at its internals much.
Is it ok for you?